extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java

```diff
@@ -169,36 +182,35 @@ public static void write(
             @NotNull final ParquetInstructions writeInstructions,
             @NotNull final String destPathName,
             @NotNull final Map<String, String> incomingMeta,
-            @NotNull final Function<String, String> groupingPathFactory,
-            @NotNull final String... groupingColumns) throws SchemaMappingException, IOException {
+            final Map<String, GroupingColumnWritingInfo> groupingColumnsWritingInfoMap)
+            throws SchemaMappingException, IOException {
         final TableInfo.Builder tableInfoBuilder = TableInfo.builder();
-        ArrayList<String> cleanupPaths = null;
+        ArrayList<File> cleanupFiles = null;
         try {
-            if (groupingColumns.length > 0) {
-                cleanupPaths = new ArrayList<>(groupingColumns.length);
-                final Table[] auxiliaryTables = Arrays.stream(groupingColumns)
-                        .map(columnName -> groupingAsTable(t, columnName))
-                        .toArray(Table[]::new);
-
+            if (groupingColumnsWritingInfoMap != null) {
+                cleanupFiles = new ArrayList<>(groupingColumnsWritingInfoMap.size());
                 final Path destDirPath = Paths.get(destPathName).getParent();
-                for (int gci = 0; gci < auxiliaryTables.length; ++gci) {
-                    final String parquetColumnName =
-                            writeInstructions.getParquetColumnNameFromColumnNameOrDefault(groupingColumns[gci]);
-                    final String groupingPath = groupingPathFactory.apply(parquetColumnName);
-                    cleanupPaths.add(groupingPath);
+                for (Map.Entry<String, GroupingColumnWritingInfo> entry : groupingColumnsWritingInfoMap.entrySet()) {
+                    final String groupingColumnName = entry.getKey();
+                    final Table auxiliaryTable = groupingAsTable(t, groupingColumnName);
+                    final String parquetColumnName = entry.getValue().parquetColumnName;
```

Comment:
General thing to think about:
`ParquetWriteInstructions` are used at multiple levels. This code could do with a cleanup around control flow and the responsibilities of each layer.